wdns/str_to_rdata_ubuf.c

```diff
@@ -197,31 +197,65 @@ str_to_svcparam(ubuf *u, char *keyval)
 		} while ((tok = strtok_r(NULL, ",", &endp)) != NULL);
 		break;
 
-	case spr_alpn:
+	case spr_alpn: {
 		/*
 		 * The wire format value for "alpn" consists of at least one
 		 * "alpn-id" prefixed by its length as a single octet, and
 		 * these length-value pairs are concatenated to form the
 		 * SvcParamValue. These pairs MUST exactly fill the
 		 * SvcParamValue; otherwise, the SvcParamValue is malformed.
 		 */
-		tok = strtok_r(NULL, ",", &endp);
+		char *src;
+		int len;
+		ubuf *v;
+		uint8_t c, bytes;
+
+		tok = strtok_r(NULL, " \t\r\n", &endp);
 		if (tok == NULL || *tok == '\0') {
 			return (wdns_res_parse_error);
 		}
 
-		do {
-			tok_len = strlen(tok);
-			if (tok_len > UINT8_MAX) {
-				return (wdns_res_parse_error);
+		src = strdup(tok);
+		len = strlen(src);
+
+		v = ubuf_new();
+		bytes = 0;
+
+		for (int i = 0; i < len; i++) {
+			c = src[i];
+
+			if (c == '\\') {
+				/* shouldn't happen, but just to be safe.. */
+				if (++i >= len) {
+					return (wdns_res_parse_error);
+				}
+				c = src[i];
+				ubuf_append(v, &c, 1);
+				bytes++;
+			} else if (c == ',') {
+				ubuf_append(u, (uint8_t *)&bytes,
```

Comment:
What happens if a comma is the first character we see? We'll append 0 to u, etc...
Maybe wrap a `if (bytes > 0` around lines 237-241?